network/yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java

```diff
@@ -100,11 +127,34 @@ private boolean isAuthenticationEnabled() {
    */
   @Override
   protected void serviceInit(Configuration conf) {
+
+    // In case this NM was killed while there were running spark applications, we need to restore
+    // lost state for the existing executors.  We look for an existing file in the NM's local dirs.
+    // If we don't find one, then we choose a file to use to save the state next time.  However, we
+    // do *not* immediately register all the executors in that file, just in case the application
+    // was terminated while the NM was restarting.  We wait until yarn tells the service about the
+    // app again via #initializeApplication, so we know its still running.  That is important
+    // for preventing a leak where the app data would stick around *forever*.  This does leave
+    // a small race -- if the NM restarts *again*, after only some of the existing apps have been
+    // re-registered, the info of the remaining apps is lost.
+    registeredExecutorFile =
+      findRegisteredExecutorFile(conf.getStrings("yarn.nodemanager.local-dirs"));
```

Comment:
yes, precisely.  I suppose I could use reflection to see if `getRecoveryPath` is available, and if not, then fall back on this mechanism.  But since I want this to work w/out `getRecoveryPath` in any case, is there any advantage to using it when its available?  maybe its a little safer when its around?